File: packages/spawn-engine-js/src/engine.mjs
// Purrden spawn engine — JS reference implementation (pure, zero-build ESM).
//
// Determinism contract (must stay byte-identical to the Python port):
//   - all scoring is signed INTEGER milli-logits (no floats)
//   - weights come from the committed integer exp LUT (content/exp_lut.json), never exp() at runtime
//   - the RNG draw is HMAC-SHA256 -> uint64 stream -> rejection sampling (no modulo bias)
//   - candidates are walked in stable content-id order
//
// `hmac(secretBytes, messageString) -> Uint8Array(32)` is injected so this file stays runtime-neutral
// (Node passes a node:crypto impl; the browser will pass a Web Crypto impl in Phase 1).

const TIERS = ["common", "uncommon", "rare", "legendary"];
const AFFINITY_KINDS = ["precipitation", "daylight", "season", "moon"];

function aff(cat, kind, value) {
  if (value == null) return 0;
  const table = cat.affinities?.[kind];
  return (table && Object.prototype.hasOwnProperty.call(table, value)) ? table[value] : 0;
}

function clampCap(value, cap) {
  return value > cap ? cap : value;
}

// Returns { score, factors:[{label,value}], pityApplied }
function scoreCat(cat, ctx, ruleset) {
  const factors = [];
  let score = cat.base_score;

  for (const kind of AFFINITY_KINDS) {
    const v = aff(cat, kind, ctx.world?.[kind]);
    if (v !== 0) factors.push({ label: `${kind}:${ctx.world[kind]}`, value: v });
    score += v;
  }
  for (const plant of ctx.placements ?? []) {
    const v = (cat.affinities?.placements && Object.prototype.hasOwnProperty.call(cat.affinities.placements, plant))
      ? cat.affinities.placements[plant] : 0;
    if (v !== 0) factors.push({ label: `plant:${plant}`, value: v });
    score += v;
  }

  // Streak raises luck for non-common cats only (a constant added to every cat cancels after max-subtract).
  if (cat.rarity !== "common") {
    const streak = clampCap((ctx.activity?.streak_days ?? 0) * ruleset.streak_bonus_per_day, ruleset.streak_bonus_cap);
    if (streak > 0) factors.push({ label: `streak:${ctx.activity.streak_days}d`, value: streak });
    score += streak;
  }

  const p = ruleset.pity[cat.rarity] ?? { per_miss: 0, cap: 0 };
  const pityBonus = clampCap((ctx.pity?.[cat.rarity] ?? 0) * p.per_miss, p.cap);
  if (pityBonus > 0) factors.push({ label: `pity:${cat.rarity}`, value: pityBonus });
  score += pityBonus;

  if (!(ctx.discovered ?? []).includes(cat.id)) {
    factors.push({ label: "discovery", value: ruleset.discovery_bonus });
    score += ruleset.discovery_bonus;
  }
  if ((ctx.recent_cats ?? []).includes(cat.id)) score -= ruleset.recent_visit_penalty;
  if ((ctx.fully_evolved ?? []).includes(cat.id)) score -= ruleset.fully_evolved_penalty;

  return { score, factors, pityApplied: pityBonus > 0 };
}

function biomeMatches(cat, ctx) {
  return cat.biome == null || cat.biome === ctx.biome;
}

// Hard gates. Fallback cats bypass level/biome so the pool is never empty.
function passesGates(cat, ctx) {
  if (!cat.enabled) return false;
  if ((ctx.cooldown_cats ?? []).includes(cat.id)) return false;
  if (cat.fallback) return true;
  return ctx.garden_level >= cat.min_garden_level && biomeMatches(cat, ctx);
}

function weightForDiff(diffMilli, lut) {
  // diffMilli <= 0. index = round(-diff / step), clamped to the table.
  const negDiff = -diffMilli;
  let idx = Math.floor((negDiff + Math.floor(lut.step_millilogits / 2)) / lut.step_millilogits);
  if (idx > lut.max_index) idx = lut.max_index;
  return lut.weights[idx];
}

function canonicalSeed(ctx, ruleset) {
  // Manual canonical serialization (sorted keys, no whitespace). Ids are simple slugs — a full
  // JCS/RFC-8785 escaper comes with real content in Phase 1. ponytail: slug-only for now.
  return '{"algorithmVersion":"' + ruleset.algorithm_version +
    '","rulesetVersion":"' + ctx.ruleset_version +
    '","slotIndex":' + ctx.slot_index +
    ',"spawnGeneration":' + ctx.spawn_generation +
    ',"spawnWindowId":"' + ctx.spawn_window_id +
    '","userId":"' + ctx.user_id + '"}';
}

function readUint64BE(bytes) {
  let v = 0n;
  for (let i = 0; i < 8; i++) v = (v << 8n) | BigInt(bytes[i]);
  return v;
}

function unbiasedDraw(hmac, secretBytes, canonical, total) {
  const T = BigInt(total);
  const TWO64 = 1n << 64n;
  const limit = (TWO64 / T) * T; // largest multiple of T that fits in 64 bits
  for (let c = 0; c < 1000; c++) {
    const digest = hmac(secretBytes, canonical + "|" + c);
    const v = readUint64BE(digest);
    if (v < limit) return Number(v % T);
  }
  throw new Error("rejection sampling failed to converge"); // astronomically unreachable
}

export function resolveSpawn(ctx, content, hmac, secretBytes) {
  const { cats, ruleset, lut } = content;

  const pityAfterInit = () => {
    const o = {};
    for (const t of TIERS) o[t] = ctx.pity?.[t] ?? 0;
    return o;
  };

  if ((ctx.occupied_slots ?? []).length >= ruleset.garden_slot_count) {
    return {
      selected: null, no_spawn_reason: "no_free_slot", fallback_used: false,
      total_weight: 0, draw: 0, eligible: [], pity_after: pityAfterInit(),
      explanation: { selected: null, tier: null, pity_applied: false, top_factors: [] },
    };
  }

  const gated = cats.filter((c) => passesGates(c, ctx));
  const fallbackUsed = !gated.some((c) => !c.fallback);

  // Score, then quantize relative to the top score.
  const scored = gated.map((c) => ({ cat: c, ...scoreCat(c, ctx, ruleset) }));
  const maxScore = Math.max(...scored.map((s) => s.score));
  const eligible = scored
    .map((s) => {
      const weight = weightForDiff(s.score - maxScore, lut);
      return { id: s.cat.id, tier: s.cat.rarity, score: s.score, weight, factors: s.factors, pityApplied: s.pityApplied };
    })
    .sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));

  const total = eligible.reduce((acc, e) => acc + e.weight, 0);
  for (const e of eligible) e.ppm = Math.floor((e.weight * 1_000_000) / total); // weight<=1e9 -> <2^53, exact

  const draw = unbiasedDraw(hmac, secretBytes, canonicalSeed(ctx, ruleset), total);

  let cumulative = 0;
  let selected = eligible[eligible.length - 1];
  for (const e of eligible) {
    cumulative += e.weight;
    if (draw < cumulative) { selected = e; break; }
  }

  const pityAfter = pityAfterInit();
  for (const t of TIERS) pityAfter[t] = t === selected.tier ? 0 : pityAfter[t] + 1;

  const topFactors = [...selected.factors]
    .filter((f) => f.value > 0)
    .sort((a, b) => (b.value - a.value) || (a.label < b.label ? -1 : 1))
    .slice(0, 3)
    .map((f) => `${f.label}(+${f.value})`);

  return {
    selected: selected.id,
    no_spawn_reason: null,
    fallback_used: fallbackUsed,
    total_weight: total,
    draw,
    eligible: eligible.map(({ id, tier, score, weight, ppm }) => ({ id, tier, score, weight, ppm })),
    pity_after: pityAfter,
    explanation: { selected: selected.id, tier: selected.tier, pity_applied: selected.pityApplied, top_factors: topFactors },
  };
}

File: packages/spawn-engine-py/purrden_spawn/engine.py
"""Purrden spawn engine — Python authoritative implementation.

Must stay byte-identical in results to the JS reference (packages/spawn-engine-js). All scoring is
integer milli-logits; weights come from the committed integer exp LUT; the draw is
HMAC-SHA256 -> uint64 stream -> rejection sampling. See engine.mjs for the shared contract.
"""
from __future__ import annotations

import hashlib
import hmac as _hmac
from typing import Any, Callable

TIERS = ("common", "uncommon", "rare", "legendary")
AFFINITY_KINDS = ("precipitation", "daylight", "season", "moon")

HmacFn = Callable[[bytes, str], bytes]


def _default_hmac(secret: bytes, message: str) -> bytes:
    return _hmac.new(secret, message.encode("utf-8"), hashlib.sha256).digest()


def _aff(cat: dict, kind: str, value: Any) -> int:
    if value is None:
        return 0
    return cat.get("affinities", {}).get(kind, {}).get(value, 0)


def _clamp_cap(value: int, cap: int) -> int:
    return cap if value > cap else value


def _score_cat(cat: dict, ctx: dict, ruleset: dict) -> dict:
    factors: list[dict] = []
    score = cat["base_score"]

    world = ctx.get("world", {})
    for kind in AFFINITY_KINDS:
        v = _aff(cat, kind, world.get(kind))
        if v != 0:
            factors.append({"label": f"{kind}:{world[kind]}", "value": v})
        score += v

    placements = cat.get("affinities", {}).get("placements", {})
    for plant in ctx.get("placements", []):
        v = placements.get(plant, 0)
        if v != 0:
            factors.append({"label": f"plant:{plant}", "value": v})
        score += v

    if cat["rarity"] != "common":
        streak = _clamp_cap(ctx.get("activity", {}).get("streak_days", 0) * ruleset["streak_bonus_per_day"],
                            ruleset["streak_bonus_cap"])
        if streak > 0:
            factors.append({"label": f"streak:{ctx['activity']['streak_days']}d", "value": streak})
        score += streak

    p = ruleset["pity"].get(cat["rarity"], {"per_miss": 0, "cap": 0})
    pity_bonus = _clamp_cap(ctx.get("pity", {}).get(cat["rarity"], 0) * p["per_miss"], p["cap"])
    if pity_bonus > 0:
        factors.append({"label": f"pity:{cat['rarity']}", "value": pity_bonus})
    score += pity_bonus

    if cat["id"] not in ctx.get("discovered", []):
        factors.append({"label": "discovery", "value": ruleset["discovery_bonus"]})
        score += ruleset["discovery_bonus"]
    if cat["id"] in ctx.get("recent_cats", []):
        score -= ruleset["recent_visit_penalty"]
    if cat["id"] in ctx.get("fully_evolved", []):
        score -= ruleset["fully_evolved_penalty"]

    return {"score": score, "factors": factors, "pity_applied": pity_bonus > 0}


def _biome_matches(cat: dict, ctx: dict) -> bool:
    return cat["biome"] is None or cat["biome"] == ctx.get("biome")


def _passes_gates(cat: dict, ctx: dict) -> bool:
    if not cat["enabled"]:
        return False
    if cat["id"] in ctx.get("cooldown_cats", []):
        return False
    if cat["fallback"]:
        return True
    return ctx["garden_level"] >= cat["min_garden_level"] and _biome_matches(cat, ctx)


def _weight_for_diff(diff_milli: int, lut: dict) -> int:
    neg = -diff_milli
    idx = (neg + lut["step_millilogits"] // 2) // lut["step_millilogits"]
    if idx > lut["max_index"]:
        idx = lut["max_index"]
    return lut["weights"][idx]


def _canonical_seed(ctx: dict, ruleset: dict) -> str:
    return (
        '{"algorithmVersion":"' + ruleset["algorithm_version"]
        + '","rulesetVersion":"' + ctx["ruleset_version"]
        + '","slotIndex":' + str(ctx["slot_index"])
        + ',"spawnGeneration":' + str(ctx["spawn_generation"])
        + ',"spawnWindowId":"' + ctx["spawn_window_id"]
        + '","userId":"' + ctx["user_id"] + '"}'
    )


def _unbiased_draw(hmac_fn: HmacFn, secret: bytes, canonical: str, total: int) -> int:
    two64 = 1 << 64
    limit = (two64 // total) * total
    for c in range(1000):
        digest = hmac_fn(secret, canonical + "|" + str(c))
        v = int.from_bytes(digest[:8], "big")
        if v < limit:
            return v % total
    raise RuntimeError("rejection sampling failed to converge")


def resolve_spawn(ctx: dict, content: dict, secret: bytes, hmac_fn: HmacFn | None = None) -> dict:
    hmac_fn = hmac_fn or _default_hmac
    cats, ruleset, lut = content["cats"], content["ruleset"], content["lut"]

    def pity_after_init() -> dict:
        return {t: ctx.get("pity", {}).get(t, 0) for t in TIERS}

    if len(ctx.get("occupied_slots", [])) >= ruleset["garden_slot_count"]:
        return {
            "selected": None, "no_spawn_reason": "no_free_slot", "fallback_used": False,
            "total_weight": 0, "draw": 0, "eligible": [], "pity_after": pity_after_init(),
            "explanation": {"selected": None, "tier": None, "pity_applied": False, "top_factors": []},
        }

    gated = [c for c in cats if _passes_gates(c, ctx)]
    fallback_used = not any(not c["fallback"] for c in gated)

    scored = [{"cat": c, **_score_cat(c, ctx, ruleset)} for c in gated]
    max_score = max(s["score"] for s in scored)
    eligible = sorted(
        (
            {
                "id": s["cat"]["id"], "tier": s["cat"]["rarity"], "score": s["score"],
                "weight": _weight_for_diff(s["score"] - max_score, lut),
                "factors": s["factors"], "pity_applied": s["pity_applied"],
            }
            for s in scored
        ),
        key=lambda e: e["id"],
    )

    total = sum(e["weight"] for e in eligible)
    for e in eligible:
        e["ppm"] = (e["weight"] * 1_000_000) // total

    draw = _unbiased_draw(hmac_fn, secret, _canonical_seed(ctx, ruleset), total)

    cumulative = 0
    selected = eligible[-1]
    for e in eligible:
        cumulative += e["weight"]
        if draw < cumulative:
            selected = e
            break

    pity_after = pity_after_init()
    for t in TIERS:
        pity_after[t] = 0 if t == selected["tier"] else pity_after[t] + 1

    top_factors = [
        f"{f['label']}(+{f['value']})"
        for f in sorted(
            (f for f in selected["factors"] if f["value"] > 0),
            key=lambda f: (-f["value"], f["label"]),
        )[:3]
    ]

    return {
        "selected": selected["id"],
        "no_spawn_reason": None,
        "fallback_used": fallback_used,
        "total_weight": total,
        "draw": draw,
        "eligible": [{k: e[k] for k in ("id", "tier", "score", "weight", "ppm")} for e in eligible],
        "pity_after": pity_after,
        "explanation": {
            "selected": selected["id"], "tier": selected["tier"],
            "pity_applied": selected["pity_applied"], "top_factors": top_factors,
        },
    }
